include/jemalloc/internal/cache_bin.h

```diff
@@ -174,10 +180,30 @@ cache_bin_nonfast_aligned(const void *ptr) {
 	return ((uintptr_t)ptr & san_cache_bin_nonfast_mask) == 0;
 }
 
+static inline const void *
+cache_bin_disabled_bin_stack(void) {
+	return &disabled_bin;
+}
+
+/*
+ * If a cache bin was zero initialized (either because it lives in static or
+ * thread-local storage, or was memset to 0), this function indicates whether or
+ * not cache_bin_init was called on it.
+ */
+static inline bool
+cache_bin_still_zero_initialized(cache_bin_t *bin) {
+	return bin->stack_head == NULL;
+}
+
+static inline bool
+cache_bin_disabled(cache_bin_t *bin) {
+	return bin->stack_head == cache_bin_disabled_bin_stack();
+}
+
 /* Returns ncached_max: Upper limit on ncached. */
 static inline cache_bin_sz_t
-cache_bin_info_ncached_max(cache_bin_info_t *info) {
-	return info->ncached_max;
+cache_bin_info_ncached_max(cache_bin_t *bin, cache_bin_info_t *info) {
+	return cache_bin_disabled(bin) ? 0: info->ncached_max;
```

Comment:
Is the cache_bin_disabled check only necessary for low_bound_get? If so, let's add the check to the corresponding top-level call site -- I'm guessing the stats gather code? If so check the bin_disabled and skip, instead of letting the logic tolerate that use case, since we have the explicit tracking of the enabled status now).

Adding the check to here means, changing the function to reading ncached_max to reading _effective_ ncached_max. I'd still say, decouple ncached_max with the enabled state -- they can sanity check each other, but not used to determine each other. For example, we should add assert(!cache_bin_disabled() here, and let that help us catch which cases go down here with disabled bins, then guard that use case at top level with the bin_disabled check.